File: 70. Climbing Stairs/ClimbStairs.py
'''
https://leetcode.com/problems/climbing-stairs/
'''

class Solution:
    def climbStairs(self, n: int):
        # naive solution+
        def findNumOfStrairs(rem, tot):
            if tot == n:
                return 1
            
            if tot > n:
                return 0

            return findNumOfStrairs(rem - 1, tot + 1) + findNumOfStrairs(rem - 2, tot + 2)

        return findNumOfStrairs(n, 0)

    
    def climbStairs2(self, n: int):
        # efficent algorithm using memoization
        memo = dict()
        def findNumOfStairs(rem, tot):
            if tot == n:
                return 1

            if tot > n:
                return 0

            if not memo[rem - 1]:
                memo[rem - 1] = findNumOfStairs(rem - 1, tot - 1)

            if not memo[rem - 2]:
                memo[rem - 2] = findNumOfStairs(rem - 2, tot - 2)

            return memo[rem - 1] + memo[rem - 2]

        return findNumOfStairs(n, 0)

    
    # iterative approach, the sequence is the same as the fibonacci numbers from my observation
    def climbStairs3(self, n: int):
        if n < 3:
            return n
        
        a_1 = 1
        a_2 = 2
        curr = a_1 + a_2
        for i in range(3, n+1):
            curr = a_1 + a_2
            a_1 = a_2
            a_2 = curr
            
        return curr

File: 70. Climbing Stairs/climb_stairs.rs
impl Solution {
    
    // O(n) time,
    // O(n) space,
    // Approach: dynamic programming, array
    pub fn climb_stairs(n: i32) -> i32 {
        if n < 2 {
            return 1;
        }
        
        let n = n as usize;
        let mut dp = vec![0;n];
        dp[n-1] = 1;
        dp[n-2] = 2;
        
        for i in (0..n-2).rev() {
            dp[i] = dp[i+1] + dp[i+2];
        }
        
        return dp[0];
    }


    // O(n) time,
    // O(1) space,
    // Approach: dynamic programming, bottom up, 
    pub fn climb_stairs2(n: i32) -> i32 {
        
        if n < 2 {
            return n;
        }
        
        let mut curr = 2;
        let mut prev = 1;
        
        for i in 0..n-2 {
            
            let temp = curr;
            curr += prev;
            prev = temp;   
        }
        
        return curr;
    }
}